ATEC-9902: Embed version metadata in addons on all platforms

vhorvath-gs opened the pull request:
With this change, the add-on tools will rely more on information provided by the add-on template's config.json file for improved UX.

Add-ons on both platforms will get their version information from the config.json file's "version" field, which must be a 1, 2 or 3 component version number in the form on "123", "1.23" or "1.2.3". Components that aren't provided will default to 0, i.e. "123.0.0" with a one component version number.

The "copyright" field of config.json is an object with fields "name" and "year", which are used to fill in company and copyright metadata where applicable. The "year" field is a format string that will be used as an input to the string(TIMESTAMP) CMake command. For details on what format options are available, see the documentation:

https://cmake.org/cmake/help/latest/command/string.html#timestamp

These new "version" and "copyright" fields are required.

Some platform specific information:

For Windows: There was previously no VERSIONINFO resource generated for addons, so this is completely new.

For MacOS: Some property list items were removed, replaced and moved around to better reflect the documentation from Apple and to get rid of obsolete and unused items.

vhorvath-gs committed:
Embed version metadata in addons on all platforms

With this change, the add-on tools will rely more on information
provided by the add-on template's config.json file for improved UX.

Add-ons on both platforms will get their version information from the
config.json file's "version" field, which must be a 1, 2 or 3 component
version number in the form on "123", "1.23" or "1.2.3". Components that
aren't provided will default to 0, i.e. "123.0.0" with a one component
version number.

The "copyright" field of config.json is an object with fields "name" and
"year", which are used to fill in company and copyright metadata where
applicable. The "year" field is a format string that will be used as an
input to the string(TIMESTAMP) CMake command. For details on what format
options are available, see the documentation:

https://cmake.org/cmake/help/latest/command/string.html#timestamp

These new "version" and "copyright" fields are required.

Some platform specific information:

For Windows: There was previously no VERSIONINFO resource generated for
addons, so this is completely new.

For MacOS: Some property list items were removed, replaced and moved
around to better reflect the documentation from Apple and to get rid of
obsolete and unused items.

vhorvath-gs commented on AddOnInfo.plist.in:
```diff
@@ -1,34 +1,29 @@
 		<key>CFBundlePackageType</key>
 		<string>.APX</string>
-		<key>CFBundleSignature</key>
-		<string>GSAP</string>
```
I have removed `CFBundleSignature`, `CFBundleLongVersionString` and `CFBundleGetInfoString` as these are obsolete and unused.

akossomorjai replied:
How do these match the XCode settings? If the developer creates an Xcode project and builds from there (e.g., because of code signing), do these behave "well"?

vhorvath-gs replied:
With the company wide Xcode version I found these:

- `CFBundleSignature`: corresponds to `Bundle creator OS Type code`
- `CFBundleLongVersionString` does not exist
- `CFBundleGetInfoString`: corresponds to `Get Info string`

These appear to be vestigial similar to the MS-DOS stub program in every Windows executable :)

akossomorjai replied:
OK, thanks!
Have you thought about adding NSHumanReadableCopyright?

vhorvath-gs replied:
That's already there and now it uses the information provided by the user via the `config.json`.

vhorvath-gs commented on BuildAddOn.py:
```diff
@@ -200,25 +200,23 @@ def GetToolset (version):
     projGenParams.append (f'-DAC_VERSION={version}')
-    projGenParams.append (f'-DAC_ADDON_NAME={addOnName}')
     projGenParams.append (f'-DAC_API_DEVKIT_DIR={str (devKitFolder / "Support")}')
-    projGenParams.append (f'-DAC_ADDON_LANGUAGE={languageCode}')
```
These are now derived from the config.json file in CMake.

vhorvath-gs replied:
Oops, not the language one actually, since this is needed to build all languages in bulk via this script.

vhorvath-gs commented on CMakeCommon.cmake:
```diff
@@ -99,10 +99,80 @@ function (LinkGSLibrariesToProject target acVersion devKitDir)
+function (generate_add_on_version_info)
+    parse_version ("${addOnVersion}" vers)
+    if (NOT DEFINED vers)
+        message (FATAL_ERROR "'${addOnVersion}' does not follow the '123' or '1.23' or '1.2.3' version format.")
```
This error message will force users to take a look at config.json and change the version there, because in the add-on template I plan to make its value `change me`. The acceptance criteria asked for no default value and this achieves that.

vhorvath-gs commented on CMakeCommon.cmake:
```diff
@@ -99,10 +99,80 @@ function (LinkGSLibrariesToProject target acVersion devKitDir)
+        math (EXPR combined "${acVersion} * 10000 + ${gsBuildNum}")
+        list (APPEND vers "${combined}")
+        list (JOIN vers . longVersion)
+        if (NOT AC_ADDON_FOR_DISTRIBUTION)
+            string (APPEND longVersion -private)
+        endif ()
```
I didn't have a better idea to mark a build as private on MacOS. This version is not the one that is displayed when a user brings up the context menu for the file and selects `Get Info`.

This can checked by Archicad later and maybe warn that there are private addons loaded. I'm just thinking of potential uses for this information. Right now there are no semantics attached to this.

akossomorjai replied:
There is already a `GSPrivateBuild` key with value 1 in the Info.plist file of the add-ons we build during the normal jam-based process. Can you please try to use that?

vhorvath-gs committed:
Support building all languages in bulk

vhorvath-gs committed:
Shift AC version left another digit

This allows the GS build num to be at most 5 digits before it overlaps.

vhorvath-gs committed:
Enable marking the build as "for dsitribution"

The "BuildAddOn.py" script now has a new command line option that will
pass "-DAC_ADDON_FOR_DISTRIBUTION=ON" to the build(s). This will prevent
the add-on from being marked as "private".

vhorvath-gs committed:
Embed AC version(s) in root product version

Since the root versions require their components to be in the 16-bit
unsigned integer range (0-65535), I have decided to move the AC version
and the GS build number into the root product version. The root product
version is not displayed in the Detail tab in Properties if you rename
the add-on to .dll, because the "StringFileInfo.ProductVersion" value
takes precedence.

Making the AC version the 2nd and the GS build number the 4th component
will make it very easy to read those versions from Archicad later. See
the "dwProductVersionMS" and "dwProductVersionLS" members:

https://learn.microsoft.com/en-us/windows/win32/api/verrsrc/ns-verrsrc-vs_fixedfileinfo

akossomorjai commented on CMakeCommon.cmake:
```diff
@@ -289,4 +323,41 @@ function (verify_api_devkit_folder devKitPath)
+    string (JSON languagesType ERROR_VARIABLE error TYPE "${json}" languages)
+    if (error OR NOT languagesType STREQUAL "ARRAY")
+        message (FATAL_ERROR "'languages' in config.json must be an array: ${error}")
+    endif()
```
Space between `endif` and `()` ?

akossomorjai commented on CMakeCommon.cmake:
```diff
@@ -289,4 +323,41 @@ function (verify_api_devkit_folder devKitPath)
+        string (JSON language ERROR_VARIABLE error GET "${json}" "${i}")
+        if (error)
+            break ()
+        endif()
```
Here as well...

akossomorjai commented on CMakeCommon.cmake:
```diff
@@ -289,4 +323,41 @@ function (verify_api_devkit_folder devKitPath)
+        endif()
+        list (APPEND addOnLanguages "${language}")
+        math (EXPR i "${i} + 1")
+    endwhile()
```
Here as well...

vhorvath-gs committed:
Use `GSPrivateBuild` to mark private builds

Info.plist files support user-defined keys and this key has a
precedence, although it is yet to be used for warning users about
potentially leaking a build not meant for distribution.

vhorvath-gs committed:
Formatting fix in new code

vhorvath-gs committed:
Fix straggler tab indents
